neuro_extras/main.py

```diff
@@ -221,6 +252,182 @@ def image() -> None:
+        if extract:
+            # extract to tmp dir
+            file = Path(source_url.path)
+            if file.is_dir():
+                file = list(file.glob("*"))[0]
+            suffixes = file.suffixes
+            if suffixes[-2:] == [".tar", ".gz"] or suffixes[-1] == ".tgz":
+                command = "tar"
+                args = ["zxvf", str(file), "-C", str(tmp_dst_url)]
+            elif suffixes[-2:] == [".tar", ".bz2"] or suffixes[-1] in (".tbz2", ".tbz"):
+                command = "tar"
+                args = ["jxvf", str(file), "-C", str(tmp_dst_url)]
+            elif suffixes[-1] == ".tar":
+                command = "tar"
+                args = ["xvf", str(file), "-C", str(tmp_dst_url)]
+            elif suffixes[-1] == ".gz":
+                command = "gunzip"
+                args = [str(file), str(tmp_dst_url) + file.name[:-3]]
+            elif suffixes[-1] == ".zip":
+                command = "unzip"
+                args = [str(file), "-d", str(tmp_dst_url)]
+            else:
+                raise ValueError(f"Don't know how to extract file {file.name}")
+
+            click.echo(f"Running {command} {' '.join(args)}")
+            subprocess = await asyncio.create_subprocess_exec(command, *args)
+            returncode = await subprocess.wait()
+            if returncode != 0:
+                raise click.ClickException(f"Extraction failed: {subprocess.stderr}")
+            else:
+                if file.exists():
+                    # gunzip removes src after extraction, while tar - not
+                    file.unlink()
+
+        # handle upload/rsync
+        await _nonstorage_cp(source_url, destination_url)
+
+
+async def _nonstorage_cp(source_url: URL, destination_url: URL) -> None:
+    if "s3" in (source_url.scheme, destination_url.scheme):
+        command = "aws"
+        args = ["s3", "cp", str(source_url), str(destination_url)]
+        if source_url.path.endswith("/"):
+            args.insert(2, "--recursive")
+    elif "gs" in (source_url.scheme, destination_url.scheme):
+        command = "gsutil"
+        args = ["-m", "cp", "-r", str(source_url), str(destination_url)]
+    elif source_url.scheme == "" and destination_url.scheme == "":
+        command = "rsync"
+        args = [
+            "-avzh",
+            "--progress",
+            "--remove-source-files",
+            str(source_url),
+            str(destination_url),
+        ]
+    else:
+        raise ValueError("Unknown cloud provider")
+    click.echo(f"Running {command} {' '.join(args)}")
+    subprocess = await asyncio.create_subprocess_exec(command, *args)
+    returncode = await subprocess.wait()
+    if returncode != 0:
+        raise click.ClickException("Cloud copy failed")
+    elif UrlType.get_type(source_url) == UrlType.LOCAL:
+        source_path = Path(source_url.path)
+        if source_path.is_dir():
+            dir_util.remove_tree(str(source_path))
+        else:
+            source_path.unlink()
```

Comment:
what's the purpose of this branch? from my perspective, it only complicates the logic without adding real value. Let's drop it?

moreover, it's broke: if `cp` command somewhy failed (e.g., no more disk space), we'll just remove source files

Reply:
The intent is not to keep src files if ingestion was performed on a local machine. 
>moreover, it's broke: if cp command somewhy failed (e.g., no more disk space), we'll just remove source files

[Here](https://github.com/neuromation/neuro-extras/blob/an/38-ingestion-and-extraction/neuro_extras/main.py#L338) the destination_url is randomly generated, so even if *aws* / *gsutil* could skip downloaded files, it wount find them..

Reply:
note, `source_path.unlink()` will fail if `source_path` is a dir. It's safer to use `shutil`

Reply:
oh I see there's a check above